### PR TITLE
Fix some issues with `unwrap_usize` instead of `assert_usize`

### LINKED ISSUE #61337: ICE when returning mutable reference to array with const generic length
EDIT: Generalized the error.

A clean rust project with just this code:
```rust
#![feature(const_generics)]

fn f<T, const N: usize>(v: &mut [T; N]) -> &mut [T; N] { v }
/* note that everything works fine if `&mut` is replaced with just `&` */

fn main() {}
```
Produces this error:
```
    Checking rust4 v0.1.0 (/home/carado/tmp/rust4)
warning: the feature `const_generics` is incomplete and may cause the compiler to crash
 --> src/main.rs:1:12
  |
1 | #![feature(const_generics)]
  |            ^^^^^^^^^^^^^^

error: internal compiler error: src/librustc/ty/sty.rs:2311: expected constant usize, got Const {
    ty: usize,
    val: Param(
        N/#1,
    ),
}

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:637:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.25/src/backtrace/libunwind.rs:97
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.25/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:197
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   6: rustc::util::common::panic_hook
   7: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   8: std::panicking::begin_panic
   9: rustc_errors::Handler::bug
  10: rustc::util::bug::opt_span_bug_fmt::{{closure}}
  11: rustc::ty::context::tls::with_opt::{{closure}}
  12: rustc::ty::context::tls::with_context_opt
  13: rustc::ty::context::tls::with_opt
  14: rustc::util::bug::opt_span_bug_fmt
  15: rustc::util::bug::bug_fmt
  16: rustc::ty::sty::Const::unwrap_usize::{{closure}}
  17: <rustc_mir::transform::qualify_consts::HasMutInterior as rustc_mir::transform::qualify_consts::Qualif>::in_rvalue
  18: rustc_mir::transform::qualify_consts::Checker::assign
  19: <rustc_mir::transform::qualify_consts::QualifyAndPromoteConstants as rustc_mir::transform::MirPass>::run_pass
  20: rustc_mir::transform::run_passes::{{closure}}
  21: rustc_mir::transform::run_passes
  22: rustc_mir::transform::mir_validated
  23: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_validated>::compute
  24: rustc::dep_graph::graph::DepGraph::with_task_impl
  25: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::force_query
  26: rustc::ty::query::plumbing::force_from_dep_node
  27: rustc::dep_graph::graph::DepGraph::try_mark_previous_green
  28: rustc::dep_graph::graph::DepGraph::try_mark_green
  29: rustc::dep_graph::graph::DepGraph::try_mark_green_and_read
  30: rustc::ty::<impl rustc::ty::context::TyCtxt>::par_body_owners
  31: rustc::util::common::time
  32: rustc_interface::passes::analysis
  33: rustc::ty::query::__query_compute::analysis
  34: rustc::dep_graph::graph::DepGraph::with_task_impl
  35: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  36: rustc::ty::context::tls::enter_global
  37: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  38: rustc_interface::passes::create_global_ctxt::{{closure}}
  39: rustc_interface::interface::run_compiler_in_existing_thread_pool
  40: std::thread::local::LocalKey<T>::with
  41: scoped_tls::ScopedKey<T>::set
  42: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
query stack during panic:
#0 [mir_validated] processing `f`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.37.0-nightly (37d001e4d 2019-05-29) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `rust4`.

To learn more, run the command again with --verbose.

```

### LINKED ISSUE #61341: ICE: rustc crashed on const generics
Test code:
``` Rust
#![feature(const_generics)]

use std::mem::MaybeUninit;

#[repr(transparent)]
pub struct OsThreadStack<const N: usize>(MaybeUninit<[u64; N]>);

fn main() {

}
```

It produced error:
``` Rust
error: internal compiler error: src/librustc/ty/sty.rs:2311: expected constant usize, got Const {
    ty: usize,
    val: Param(
        N/#0,
    ),
}
```

Playground version: rustc 1.37.0-nightly (2019-05-29 37d001e4d)

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=dc28c72478b76d23c757772335970b68)

### LINKED ISSUE #61422: ICE: returning array with Const Generic size from mem::uninitialized()
```
error: internal compiler error: src/librustc/ty/sty.rs:2311: expected constant usize, got Const {
    ty: usize,
    val: Param(
        SIZE/#0,
    ),
}

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:637:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to previous error
```

```
#![feature(const_generics)]
use std::mem;

fn foo<const SIZE: usize>() {
    let arr: [u8; SIZE] = unsafe {
        let mut array: [u8; SIZE] = mem::uninitialized();
        
        array // remove and removes error
    };
}
```

### PR DESCRIPTION
**Pull request by varkor**
Fixes https://github.com/rust-lang/rust/issues/61337.
Fixes https://github.com/rust-lang/rust/issues/61341.
Fixes https://github.com/rust-lang/rust/issues/61422.

r? @eddyb

### CODE REVIEW

File: src/librustc_typeck/check/_match.rs [hunk 1]
```diff
@@ -400,27 +400,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
-                                    "pattern requires at least {} elements but array has {}",
-                                    min_len, size)
-                                .span_label(pat.span,
-                                    format!("pattern cannot match array of {} elements", size))
-                                .emit();
+                            struct_span_err!(
+                                tcx.sess,
+                                pat.span,
+                                E0730,
+                                "cannot pattern-match on an array without a fixed length",
```
Comment (Centril) [inline src/librustc_typeck/check/_match.rs]:
```suggestion
                                "cannot pattern-match on an array with unknown length",
```

File: src/test/ui/const-generics/issue-61422.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// run-pass
+
+#![feature(const_generics)]
+//~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash
+
+use std::mem;
+
+fn foo<const SIZE: usize>() {
+    let arr: [u8; SIZE] = unsafe {
+        let mut array: [u8; SIZE] = mem::uninitialized();
```
Comment (Centril) [inline src/test/ui/const-generics/issue-61422.rs]: We should think about how we handle this in terms of post monomorphization errors given that if you substitute `u8` for a larger type and set `SIZE` to a huge number then `foo` would error post monomorphization. Leave a comment for now? cc @oli-obk

File: src/librustc_typeck/check/_match.rs (hunk 1 above)
Comment (varkor) [reply src/librustc_typeck/check/_match.rs]: I'm not sure "unknown" is accurate here. The length is known: it's just parameterised.

File: src/test/ui/const-generics/issue-61422.rs (hunk 2 above)
Comment (varkor) [reply src/test/ui/const-generics/issue-61422.rs]: Isn't this an issue for large allocations in general? E.g. https://github.com/rust-lang/rust/issues/55878.

File: src/librustc/ty/layout.rs [hunk 3]
```diff
@@ -548,28 +548,30 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                         return Err(LayoutError::Unknown(ty));
                     }
                 }
+                if let Some(count) = count.assert_usize(tcx) {
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: You should use `count.assert_usize(tcx).ok_or(LayoutError::Unknown(ty))?` to minimize the diff.

File: src/test/ui/const-generics/issue-61422.rs (hunk 2 above)
Comment (Centril) [reply src/test/ui/const-generics/issue-61422.rs]: Yes, and we can do the same with e.g.:
```rust
fn foo<T>() {
    let _x: [T; 10000000000000];
}

fn main() {
    foo::<[u8; 1000]>();
}
```
but that doesn't necessarily mean we should introduce another avenue for post monomorphization errors.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 4]
```diff
@@ -343,7 +343,11 @@ fn fixed_vec_metadata(
 
     let upper_bound = match array_or_slice_type.sty {
         ty::Array(_, len) => {
-            len.unwrap_usize(cx.tcx) as c_longlong
+            if let Some(len) = len.assert_usize(cx.tcx) {
+                len as c_longlong
+            } else {
+                -1
+            }
```
Comment (eddyb) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: Is this panicking? This should be monomorphic by now.

File: src/librustc_codegen_ssa/mir/rvalue.rs [hunk 5]
```diff
@@ -519,8 +519,9 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         if let mir::Place::Base(mir::PlaceBase::Local(index)) = *place {
             if let LocalRef::Operand(Some(op)) = self.locals[index] {
                 if let ty::Array(_, n) = op.layout.ty.sty {
-                    let n = n.unwrap_usize(bx.cx().tcx());
-                    return bx.cx().const_usize(n);
+                    if let Some(n) = n.assert_usize(bx.cx().tcx()) {
+                        return bx.cx().const_usize(n);
+                    }
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/rvalue.rs]: Same here, there's no reason to handle generic code here.

File: src/test/ui/const-generics/issue-61422.rs (hunk 2 above)
Comment (eddyb) [reply src/test/ui/const-generics/issue-61422.rs]: Types can depend on constants and constants can depend on types - this is *not* a new avenue.

Comment (Centril) [reply src/test/ui/const-generics/issue-61422.rs]: @eddyb Even so, this increases the surface area for post monomorphization errors and in this case we can force the user to introduce a WF constraint `where [u8; N]:,`. We'll need to consider this at one time or another.

Comment (varkor) [reply src/test/ui/const-generics/issue-61422.rs]: I don't think this belongs in a test case for a random issue. If this is viewed as a separate concern, it should be opened in its own issue.

**Comment by varkor** [comment]
I've removed the unnecessary guards in the rebase.

File: src/test/ui/const-generics/issue-61422.rs (hunk 2 above)
Comment (Centril) [reply src/test/ui/const-generics/issue-61422.rs]: Fine; I wanted this as a reminder when reviewing the test suite so I don't forget about it when we near stabilization... but I'll think about an issue instead.

**Review by eddyb — APPROVED** [review]

**Pull request merged**